### ISSUE TITLE
Labirint game: sprite collision falsely detected

### ISSUE THREAD

**Issue by MBilderbeek** [openMSX/openMSX#1198]
This Russian game Labirint doesn't work in openMSX (not on any machine I tried), when starting the game, it directly detects a hit and you loose a life, until you're game over. On my real turboR, it runs fine. So something is clearly wrong in the openMSX sprite collision implementation.

The disk image contains two games, both show the issue: labirin1 and labirin2.

[labirint.dsk.gz](https://github.com/openMSX/openMSX/files/3868566/labirint.dsk.gz)

**Comment by mthuurne** [comment]
It seems you are right about the sprite collision being essential: if I disable the collision bit (S#0 bit 5) in SCREEN2, the player becomes invulnerable.

The "labirin1" game uses just 2 sprites: sprite 0 for the player and sprite 1 for the enemy. However, it doesn't disable the other sprites, so sprite 2 through 31 will be processed by the VDP. The data for these unused sprites is all zeroes, so this leads to pattern 0 (one of the player walking animations) to be displayed in the top-left corner in color 0 (invisible).

If I drop sprites with color 0 from the `spriteBuffer` in `SpriteChecker::checkSprites1()`, the problem goes away. So my guess is that the real VDP only detects collisions between pixels that have non-zero color values and treats zero-colored sprites as if their pattern has nothing but zeroes. This would have to be verified on real hardware before we implement it.

**Comment by mthuurne** [comment]
If the theory that zero-colored sprites cannot collide is correct, we should check at what point the VDP ignores collisions on them. Does it drop them before or after fetching their patterns? This can be determined by checking whether zero-colored sprites count towards the 4 sprites per line limit.

**Comment by grauw** [comment]
On V99x8 I would assume this behaviour changes depending on the TP flag?

**Comment by Wierzbowsky** [comment]
The bug is correctly identified. All sprites in these games beside 0 and 1 have transparent color (because they are not defined after initialization). When the unused sprites' color is changed, the collision problem gets resolved without any side effects. What I did in the fix (it's a hack rather than a fix for those who like to play the games now) was just writing 0xFF to the 4th byte of every sprite's entry in the Sprite Attribute Table, except for 0 and 1 sprites. So to fix the problem, the emulator must start ignoring the collision of sprites with transparent color in their SAT entry.

**Comment by m9710797** [comment]
I ran the following BASIC test program on my turboR (and on openMSX):

  10 SCREEN 2
  20 OPEN"GRP:"FOR OUTPUT AS #1
  30 PRINT#1,"test"
  50 B$="":FORI=0TO7:B$=B$+CHR$(255):NEXT
  60 SPRITE$(0)=B$
  70 PUT SPRITE 0,(100,100),0,0
  71 PUT SPRITE 1,(100,100),0,0
  72 PUT SPRITE 2,(100,100),0,0
  73 PUT SPRITE 3,(100,100),0,0
  80 PUT SPRITE 9,(102,102),1,0
  90 ON SPRITE GOSUB 200
  100 SPRITE ON
  190 GOTO 190
  200 PRINT#1,"collision"
  210 RETURN

On my real turboR machine:
- sprite 9 is partially visible  (meaning sprites 0-3 _do_ count towards the sprite limit)
- it does not print "collision"
On openMSX:
- sprite 9 is partially visible (ok)
- it wrongly prints the text "collision"

Adapting the test for screen 5 (and using 8 iso 4 sprites with color 0) behaves the same. Except when (as grauw suggested) the TP bit is set to 1, in that case sprite color 0 _is_ visible and _does_ produce collisions.

When looking at the VRAM timing results:
  https://openmsx.org/vdp-vram-timing/vdp-timing.html
it's logical that 0-colored sprites do contribute towards the 4/8 visible sprite limit, because that decision can only be based on the sprite y-coordinates.

I could only test on a V9958, I expect that a TMS99x8 will behave the same. But could someone verify this by running the above BASIC program on a real MSX1 machine? Thanks.

Next step: I'll try to correct openMSX.

**Comment by m9710797** [comment]
I implemented a fix: now the above BASIC test behaves the same in openMSX as on a real turboR. I did not verify the 'labirint' software (I don't have it). So please help test.

@mthuurne: in this commit:

   commit 9cb4fa46d79dabce4794268039a9d67054f6e7c0
   Author: Maarten ter Huurne <maarten@treewalker.org>
   Date:   Mon Dec 24 11:10:03 2001 +0000
   Added render routine for Graphic4, support for 212 lines display, partial sprite mode 2 implementation.

You added this comment:

  Colour doesn't matter: sprites of colour 0 can collide.
  TODO: V9938 data book denies this (page 98).

But that comment contradicts my fix. I know it's a _long_ time ago since you wrote that code, but do you still have any idea what that comment was based on?

**Comment by mthuurne** [comment]
Yeah, it's way too long ago to remember any details. The comment is very specific and mentions that it contradicts the data book, so it must have been based on a measurement. It's possible it was a faulty measurement, but it could also be that there really is a situation in which sprites of color 0 can collide (other than TP=1, which is mentioned on data book page 98).

**Comment by happppp** [comment]
From p_gimeno on the chat: "I've tested this code in the VG-8020/40 and it repeatedly prints "collision"."

**Comment by TheShich** [comment]
m9710797, you can get labirints here:
https://drive.google.com/file/d/1tpkKPZydeFIoS__WjXuM5_lEfYq_WLOj/view?usp=sharing

**Comment by m9710797** [comment]
(On IRC) mthuurne suggested to also try swapping the order (sprite-priority) in which a zero- and a non-zero-colored sprite collide. Here's my new test program:

100 DEFINTA-Z:DIMR(4):SCREEN2
110 'VDP(9)=VDP(9)OR32
120 VDP(9)=VDP(9)AND223
130 OPEN"GRP:"FOR OUTPUT AS #1
140 FORI=0TO7:B$=B$+CHR$(255):NEXT
150 SPRITE$(0)=B$:ONSPRITEGOSUB240
160 FORT=0TO3:CLS:C=0
170 PRESET(10,10):PRINT#1,"test";T
180 PUTSPRITE0,(100,100),(T AND1),0
190 PUTSPRITE1,(101,101),((T/2)AND1),0
200 SPRITEON:FORW=0TO1000:NEXT
210 R(T)=C:NEXT
220 SCREEN0:FORT=0TO3:PRINTR(T);:NEXT
230 END
240 SPRITEOFF:C=1
250 PRESET(10,180):PRINT#1,"collision"
260 RETURN

On a MSX1 (with TMS99x8) comment-out/remove both lines 110 and 120. On a MSX2 or higher (with V99x8) enable either line 110 (for TP=1) or line 120 (for TP=0), also test both screen 2 and screen 5 (change line 100).

On my turbor (with V9958) I got these results (same results for screen 2 and 5):

   TP=0: 0 0 0 1
   TP=1: 1 1 1 1

This confirms the results of my previous measurement:
- zero-colored pixels, with TP=0 (thus non-visible) do NOT cause collisions
- the order does not matter

It would be interesting to also run this test on a real TMS99x8 (or variants). Based on happppp's/p_gimeno's report above, I predict the result will be "1 1 1 1". Meaning that zero-colored sprites DO collide.

**Comment by mr-GreyWolf** [comment]
At both screen 2 and screen 5 (line 100),
```
110 'VDP(9)=VDP(9)OR32
120 VDP(9)=VDP(9)AND223

MSX-2  : 0 0 0 1
openMSX: 1 1 1 1

```

```
110 VDP(9)=VDP(9)OR32
120 'VDP(9)=VDP(9)AND223

MSX-2  : 1 1 1 1
openMSX: 1 1 1 1

```

**Comment by MBilderbeek** [comment]
@mr-GreyWolf Thanks! Which MSX2 was that exactly? Just curious... I guess it has a V9938, of course. More importantly: which openMSX version did you use?

**Comment by mr-GreyWolf** [comment]
@MBilderbeek

Yamaha YIS-503IIIR
openMSX 0.15.0.

**Comment by m9710797** [comment]
@mr-GreyWolf : thanks.
The openMSX MSX2 behavior was already fixed in revision 169b3200197d (2019/11/30). Today I also fixed the MSX1 behavior (that is for MSX1 revert back to the old behavior).

**Comment by MBilderbeek** [comment]
p_gimeno wrote he confirmed the MSX1 behaviour also occurs on T6950.

With that, I just assume it happens like that on all MSX1 VDP's and I'll close this ticket. Please reopen if you find a VDP where the behaviour is different :)
